Fix #5653: improve UserPermissions - password mutations

ScottAgirs opened the pull request:
<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:
- Create or update the documentation.
- Create or update the tests.
- Refer to the issue you are closing in the PR description - fix #issue
- Specify if the PR is in WIP (work in progress) state or ready to be merged
-->

### The Problem
Current reset-password endpoints are pointing to "changePassword" Auth controller which is misleading and does not actually do the **change password** action in the traditional sense, it rather does what the endpoint says - resets user's password.

### Intent of this PR
To conform and improve reset password naming and yield way for potential future implementation of an actual **change password** functionality, while at the same time making it easier and less confusing for the user to navigate and orient oneself in the password mutations.

~To enable users to be able to change their password with their current password (without the need to generate `code`).~
~Currently, unless badly overlooking something, it's only possible by resetting the password via forgot-password.~


### Description of what you did:
#### Added
- documentation for GraphQL reset password mutation
~- `change-password` mutation (replaces the previous `change-password` - which if this is merged becomes `reset-password`)~
~- GraphQL `changePassword` mutation (replaces the previous `changePassword` - which if this is merged becomes `resetPassword`)~

#### Refactored
~- **[BREAKING]** rename `change-password` to `reset-password`~
- renamed the Auth controller from `changePassword` to `resetPassword`
- renamed response payload in GraphQL from `ForgotPassword` to `UserPersmissionsPasswordPayload`

alexandrebodin commented:
Can you explain the intent of the PR in the description plz ?

ScottAgirs commented:
@alexandrebodin done, hope this makes sense

roelbeerens commented:
@ScottAgirs I like the idea of a changePassword feature! Also, thank you for providing the documentation (I kinda forgot about those). 

The only thing that worries me right now on the changePassword mutation is, that it can only by used with providing a Bearer token with the request. The general idea is that the user should be logged in to change it's password, of course. However, I do think the 'endpoint' / controller should also be available on the 'public' side (if you want to when setting the permissions). I think the mutation should have a 'code' variable/parameter available to resolve this. This however, does make the mutation pretty much redundant and a copy of the current 'changePassword' mutation, except for the current password confirmation. That would be the main difference.

Here's what I think that's missing:

- Leave the renaming as it is, I think it's more logical this way.
- Add a 'code' parameter to the changePassword mutation (the code is a JWT token anyway and should contain the current user ID)
- Instead of using the 'state.user' in the controller, decode the code and use that user information.

With this, the endpoint/controller can be used both as a 'public' user, as a 'authenticated' user.

If I overlooked something, please let me know, I didn't fully dug through your PR ;)

Great rename on the ForgotPassword payload btw!

alexandrebodin commented:
Hi, for security reasons, I really don't think a call to an api should be able to change a user password. I would rather only have the reset password workflow. Same for the administration of users. I think we should change this to only be able to send a reset link to the user.

roelbeerens commented:
@alexandrebodin You've got a point there.

ScottAgirs commented:
@roelbeerens thanks for the feedback! Was hoping you'll be ok with the ForgotPassword payload rename, haha! The token is a nice idea!

@alexandrebodin I understand, but changing password is quite a basic and important ability to have for the users, no? Or would you guys say that changing the password by the current password is a thing of the past?

ScottAgirs commented:
PS. as an extra security layer we could add email notification about the action - in case we do decide to move forward with this

alexandrebodin commented:
I think this topics needs more thinking to be put into it before take a decision about this. I would be ok with an api call that would trigger a reset password flow for now only.

ScottAgirs commented:
> I think this topics needs more thinking to be put into it before take a decision about this. I would be ok with an api call that would trigger a reset password flow for now only. This would be equivalent to the forgot password basically

That's fair. I'll adjust the PR, so that the edits are only related to the reset password flow

ScottAgirs commented:
@alexandrebodin I have cleaned up, updated the intent and defined the problem in the description. Please take a look and let me know what you think.
@roelbeerens perhaps you could chime in too?

alexandrebodin commented:
This looks good to me, this will however be a breaking change, I'm flagging it as so and will merge it after our next release

alexandrebodin commented:
Hi @ScottAgirs We have made some updates to files you ahve changed recently would you mine resolving the conflicts so we can move this PR into the next release ?

ScottAgirs commented:
> Hi @ScottAgirs We have made some updates to files you ahve changed recently would you mine resolving the conflicts so we can move this PR into the next release ?

Hi @alexandrebodin thanks for the update, just pushed

alexandrebodin commented on packages/strapi-plugin-users-permissions/config/schema.graphql.js:
```diff
@@ -53,18 +53,18 @@ module.exports = {
-    register(input: UsersPermissionsRegisterInput!): UsersPermissionsLoginPayload!
-    forgotPassword(email: String!): ForgotPassword
-    changePassword(password: String!, passwordConfirmation: String!, code: String!): UsersPermissionsLoginPayload
+    register(input: UserInput!): UsersPermissionsLoginPayload!
```
Small mistake when you fixed the conflicts :) We should keep the UsersPermissionsRegisterInput :)

alexandrebodin replied:
Hi @ScottAgirs  will you be able to make the change ?

ScottAgirs replied:
@alexandrebodin hey Alex, thanks for the follow-up, my apologies, I must have missed the notification about your review. 

Just pushed 🚀

alexandrebodin commented on packages/strapi-plugin-users-permissions/config/schema.graphql.js:
```diff
@@ -53,18 +53,18 @@ module.exports = {
-    register(input: UsersPermissionsRegisterInput!): UsersPermissionsLoginPayload!
-    forgotPassword(email: String!): ForgotPassword
-    changePassword(password: String!, passwordConfirmation: String!, code: String!): UsersPermissionsLoginPayload
+    register(input: UserInput!): UsersPermissionsRegisterInput!
```
You forgot the input too :D `UsersPermissionsRegisterInput `

ScottAgirs replied:
ahh lol ebed726d7

sanojsilva commented on packages/strapi-plugin-users-permissions/config/schema.graphql.js:
```diff
@@ -57,8 +57,8 @@ module.exports = {
     register(input: UserInput!): UsersPermissionsLoginPayload!
-    forgotPassword(email: String!): ForgotPassword
-    changePassword(password: String!, passwordConfirmation: String!, code: String!): UsersPermissionsLoginPayload
+    forgotPassword(email: String!): UserPersmissionsPasswordPayload
```
Hey there's a spelling mistake here User**Persmissions**PasswordPayload** it should be User**Permissions**PasswordPayload

alexandrebodin replied:
nice catch, at least the code works :) Mind opening a PR ?

ScottAgirs replied:
nice catch, Sanoj, hope u don't mind I had it open, so I just pushed the patch :) https://github.com/strapi/strapi/pull/6390
